### PR TITLE
Changed LoadState to not assume that the first entry is for the policy ref

### PR DESCRIPTION
**Pull request by neilnaveen**
- Created a helper function that is almost the same as GetFirstEntry, but with an expected targetRef

### CODE REVIEW

File: internal/rsl/rsl.go [hunk 1]
```diff
@@ -506,6 +506,51 @@ func GetFirstEntry(repo *git.Repository) (*ReferenceEntry, []*AnnotationEntry, e
 	return firstEntry, annotations, nil
 }
 
+// GetFirstEntryForRef returns the very first entry in the RSL for a ref. It is expected to be
+// a reference entry as the first entry in the RSL cannot be an annotation.
+func GetFirstEntryForRef(repo *git.Repository, targetRef string) (*ReferenceEntry, []*AnnotationEntry, error) {
```
Comment (adityasaky) [inline internal/rsl/rsl.go:511]: I think there's a lot of repetition with `GetFirstEntry` except for the single targetRef check. Can you update this so `GetFirstEntry` calls this one with `targetRef=""`, and only do the `targetRef == entry.RefName` check `if targetRef != ""`?

**Review by adityasaky — CHANGES_REQUESTED** [review]
Thanks! I think this is good to go once we have this wired up to `GetFirstEntry`.

File: internal/rsl/rsl.go (hunk 1 above)
Comment (neilnaveen) [reply internal/rsl/rsl.go:474]: Fixed

File: internal/rsl/rsl.go [hunk 2]
```diff
@@ -487,8 +493,10 @@ func GetFirstEntry(repo *git.Repository) (*ReferenceEntry, []*AnnotationEntry, e
 					// The first entry cannot be an annotation
 					return nil, nil, ErrInvalidRSLEntry
 				}
-				firstEntry = entry
-				break
+				if entry.RefName == targetRef || targetRef == "" {
```
Comment (adityasaky) [inline internal/rsl/rsl.go:496]: I might have got it wrong last night. Does this still work when the first entry for a ref is actually the 17th entry overall? We hit this only when there aren't any parents, right? Do we want to continually track firstEntry outside this err block?

Comment (neilnaveen) [reply internal/rsl/rsl.go]: I have fixed this, and changed the logic, the last check at the end is just if we never encounter a rsl entry

File: internal/policy/policy_test.go [hunk 3]
```diff
@@ -68,93 +68,195 @@ func TestInitializeNamespace(t *testing.T) {
...
-	}
+		loadedState, err = LoadState(context.Background(), repo, entry.(*rsl.ReferenceEntry))
+		if err != nil {
+			t.Error(err)
+		}
 
-	assert.Equal(t, state, loadedState)
+		assert.Equal(t, state, loadedState)
+	})
+	t.Run("fail loading while verifying multiple states, bad sig", func(t *testing.T) {
```
Comment (adityasaky) [inline internal/policy/policy_test.go:160]:
```suggestion

	t.Run("fail loading while verifying multiple states, bad sig", func(t *testing.T) {
```

File: internal/policy/policy_test.go [hunk 4]
```diff
@@ -68,93 +68,195 @@ func TestInitializeNamespace(t *testing.T) {
...
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		signer, err := signerverifier.NewSignerVerifierFromSecureSystemsLibFormat(rootKeyBytes) //nolint:staticcheck
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		badSigner, err := signerverifier.NewSignerVerifierFromSecureSystemsLibFormat(targets1KeyBytes) //nolint:staticcheck
```
Comment (adityasaky) [inline internal/policy/policy_test.go:196]: move this down to when it's first used, please!

File: internal/policy/policy_test.go [hunk 5]
```diff
@@ -68,93 +68,195 @@ func TestInitializeNamespace(t *testing.T) {
...
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		state.TargetsEnvelope = env
+
+		if err := state.Commit(repo, "", false); err != nil {
+			t.Fatal(err)
+		}
+		policyStagingRef, err := repo.Reference(plumbing.ReferenceName(PolicyStagingRef), true)
```
Comment (adityasaky) [inline internal/policy/policy_test.go:237]:
```suggestion

		policyStagingRef, err := repo.Reference(plumbing.ReferenceName(PolicyStagingRef), true)
```

File: internal/policy/policy_test.go [hunk 6]
```diff
@@ -68,93 +68,195 @@ func TestInitializeNamespace(t *testing.T) {
...
+		if err := rsl.NewReferenceEntry(PolicyRef, policyStagingRef.Hash()).Commit(repo, false); err != nil {
+			t.Fatal(err)
+		}
+
+		entry, err = rsl.GetLatestEntry(repo)
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		loadedState, err = LoadState(context.Background(), repo, entry.(*rsl.ReferenceEntry))
```
Comment (adityasaky) [inline internal/policy/policy_test.go]: _, err = ...

File: internal/policy/policy_test.go [hunk 7]
```diff
@@ -68,93 +68,195 @@ func TestInitializeNamespace(t *testing.T) {
...
+
```
Comment (adityasaky) [inline internal/policy/policy_test.go:257]:
```suggestion
```

File: internal/policy/policy_test.go [hunk 8]
```diff
@@ -68,93 +68,195 @@ func TestInitializeNamespace(t *testing.T) {
...
+		assert.Equal(t, "requested state has invalidly signed metadata: verifier's key and threshold constraints not met", err.Error())
```
Comment (adityasaky) [inline internal/policy/policy_test.go:258]: I know the text is from the `fmt.Errorf` but I think the embedded error is named. Can we just do assert.ErrorIs on that one? Probably some thing about verifier constraints not being met or so

**Review by adityasaky — CHANGES_REQUESTED** [review]

File: internal/policy/policy_test.go (hunk 8 above)
Comment (neilnaveen) [reply internal/policy/policy_test.go:258]: how would I be able to compare that single part of the error?

File: internal/policy/policy_test.go (hunk 3 above)
Comment (neilnaveen) [reply internal/policy/policy_test.go:161]: fixed

File: internal/policy/policy_test.go (hunk 4 above)
Comment (neilnaveen) [reply internal/policy/policy_test.go]: fixed

File: internal/policy/policy_test.go (hunk 5 above)
Comment (neilnaveen) [reply internal/policy/policy_test.go:239]: fixed

File: internal/policy/policy_test.go (hunk 6 above)
Comment (neilnaveen) [reply internal/policy/policy_test.go]: fixed

File: internal/policy/policy_test.go (hunk 7 above)
Comment (neilnaveen) [reply internal/policy/policy_test.go]: fixed

File: internal/policy/policy_test.go (hunk 8 above)
Comment (adityasaky) [reply internal/policy/policy_test.go:259]: Does assert.ErrorIs not match nested errors?

**Review by adityasaky — APPROVED** [review]

**Pull request merged**